### ISSUE TITLE
Improvement request: Branching model

### ISSUE THREAD

**Issue by drew2a** [Tribler/tribler#5569]
## TLDR

Let's use [OneFlow](https://www.endoflineblog.com/oneflow-a-git-branching-model-and-workflow) branching model.

## Current  branching model

```
Note: below I will use current release number (7.5). 
This will be done for simplification. 
You can think of it as MAJOR.MINOR in SemVer notation.
```

Current branching model could be described as:

1. We have `devel` branch for commits during developing phase
1. We have long-living `release-7.5`
1. We add tags `7.5.1`, `7.5.2`, `7.5.3` to `release-7.5` branch while we deliver the release (no merge to `devel` or `master`)
1. When we move to the next release (7.6) we will merge `release-7.5` to the `develop` and sometimes to the `master`

```
Note: current master version: 7.3.0
```

It looks confusing and counter-intuitively. 
Missed merges to `devel` from `release-7.5` (for example) and general non-strictness, will inevitably lead to bugs.

## Proposition

We must follow one of the well-known methodologies, since there is no obvious need to reinvent it.
Strict adherence to the methodology will allow avoiding many mistakes "by negligence".

Due to long release cycle of Tribler and unstable codebase (in general) I propose to use modern [OneFlow](https://www.endoflineblog.com/oneflow-a-git-branching-model-and-workflow) or more conservative [GitFlow](https://nvie.com/posts/a-successful-git-branching-model/) model, because they are more suited to our needs.

### Alternatives

* [GitFlow](https://nvie.com/posts/a-successful-git-branching-model/)
* [GitLab Flow](https://about.gitlab.com/blog/2014/09/29/gitlab-flow/)
* [GitHub flow](https://guides.github.com/introduction/flow/)

**Comment by qstokkink** [comment]
> Missed merges to devel from release-7.5 (for example) and general non-strictness, will inevitably lead to bugs.

We have historical evidence that missing commits and weird merge conflict resolutions happen and this is the 6th(-ish?) time we will change our branching model to something that will "once and for all resolve all of our issues with branching".

Personally, I'm a big proponent of keeping stuff simple. The less I have to think about as a developer, the better. Once there's a human in the loop, there's room for error. This is why in IPv8, I enforce the simplest solution of a GitHub flow. Obviously, Tribler has more going on and we _need_ something more complex. Is OneFlow the answer? Possibly. I believe it hinges on this comment:

> Strict adherence to the methodology will allow avoiding many mistakes "by negligence".

This has been our downfall so far. I believe we have yet to 100% adhere to any methodology before. I'm more interested in how OneFlow helps in having both open source contributors and seasoned developers adhere to it, than what the output of `git log --graph` is (which is what the discussion around most of these flow model descriptions turn into). This leads me to this:

> It looks confusing and counter-intuitively.

If that implies that OneFlow is more intuitive to follow than the current flow, I fully support it.

**Comment by ichorid** [comment]
Basically OneFlow is GitFlow with garbage collection of older branches and without "master" branch.

So, now we're actually using OneFlow without garbage collection.

**Comment by ichorid** [comment]
@drew2a , BTW, we're actually merging back changes from `release-x` to `devel`. Look more closely at our git history.

**Comment by ichorid** [comment]
And yeah, no one cares about `master` branch for a long time now :shrug:

**Comment by drew2a** [comment]
@ichorid 
> BTW, we're actually merging back changes from release-x to devel. Look more closely at our git history
That's true. But but not always. As an example see release `7.5.2`.

Release branch should'n leave for a long time (how long `release 7.5` has been around?)

From my point of view you should start `release` branch from `develop`. 
Then stable it. Then close it, and merge back.

Very rarely, in extreme cases, you should make a `hotfix` from previously released branch.

**Comment by ichorid** [comment]
> From my point of view you should start `release` branch from `develop` (or master).
> Then stable it. Then close it, and merge back.

`devel` is often full of half-implemented features. We have a policy that we do not tolerate dead code in the repo, because of all the student coding patterns... So, developers who are working on any feature that takes more than a week of work are forced to either suffer rebase hell or commit half-working experimental stuff directly into `devel` (which obviously leaves `devel` in bad shape). That's why it takes so long to stabilise the release branch.

To address this problem, we should change our rules to tolerate dead/disabled code in the codebase, probably by using [Feature Flags](https://github.com/Tribler/tribler/issues/5558) and at least some project management.

**Comment by ichorid** [comment]
Also, we had :hankey: :boom: :hankey: instead of software architecture before. The code was completely unmanageable. That's another reason why we could not afford long-term project planning.

**Comment by synctext** [comment]
Some code has been experimental for [10 years, now getting usable](https://www.tribler.org/IPv8/). Now we hopefully have the luxury of getting the codebase clean everywhere. OneFlow sounds cool and simple, but I don't have operational rebase experience to judge this. Do we want the purist flavor ```Option #3 – rebase + merge –no–ff```?

Instead of using feature flags I'm in favor of using a variant on the startup scene approach with Minimal Viable Products. An **MVC** is a Minimal Viable Community with only part of the desired functionality, but can be deployed in the wild and is designed to be future-proof. The MVC approach in combination with monthly release cycle forces us to make baby steps and stability is always 1 revert away.

**Comment by ichorid** [comment]
You mean "revert incomplete community when forking for a release branch"? If we do this, we will not be able to merge back changes from the release branch back to devel branch, because these changes will contain the revert commit.

**Comment by drew2a** [comment]
@ichorid MVC is make sense :)
**Rule**: if you merge something to `devel` this should mean that you are confident that the code works, and you want this code to be included to a next release. 

Now the question is "What will you do if this code fails?". The usual answer is "You should fix it".
If your code is not meet a criteria for MVC, then don't merge it to `devel`.

@synctext I have always preferred `Option #2 – merge –no-ff` because I use GUI for git graph visualization, and for me `v2` seem more informative and tidy (a half of developers would disagree with me).

v3 looks like "advanced v2", so why not.
it is up to dev team :)

**Comment by drew2a** [comment]
Let's go further with this improvement request.

To apply it, we need to:
1. Merge `devel` branch to `master`
1. Remove `devel` branch
1. Reconfigure all Jenkins jobs
1. Describe naming rules

I can accomplish all these tasks.

What did I miss with the migration to the master?

**Comment by qstokkink** [comment]
Shouldn't it be `main` instead of `master` with the new GitHub policies?

**Comment by drew2a** [comment]
Yes. I missed it: https://github.com/github/renaming

**Comment by drew2a** [comment]
Let's postpone it to January 2021.

> Beginning in January 2021, we'll begin rolling out functionality to make it easier to rename the default branch of an existing repository.

https://github.com/github/renaming#creating-a-seamless-move-for-existing-repositories-
